iceoryx_utils/include/iceoryx_utils/internal/lockfree_queue/index_queue.hpp

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "iceoryx_utils/internal/lockfree_queue/buffer.hpp"
+#include "iceoryx_utils/internal/lockfree_queue/cyclic_index.hpp"
+#include "iceoryx_utils/internal/lockfree_queue/unique.hpp"
+
+#include <atomic>
+
+namespace iox
+{
+/// @brief lockfree queue capable of storing indices 0,1,... Capacity-1
+template <uint64_t Capacity, typename ValueType = uint64_t>
+class IndexQueue
+{
+  public:
+    class UniqueIndex : public unique<ValueType>
+    {
+      public:
+        using value_t = ValueType;
+        using invalid_t = typename unique<ValueType>::invalid_t;
+        static constexpr invalid_t invalid{};
+
+        friend class IndexQueue<Capacity, ValueType>;
+
+        // only an invalid index can be constructed by anyone other than the IndexQueue itself
+        UniqueIndex(invalid_t)
+            : unique<ValueType>(invalid)
+        {
+        }
+
+      private:
+        // valid construction is made private and only accessible by the IndexQueue
+        UniqueIndex(const ValueType& value)
+            : unique<ValueType>(value)
+        {
+        }
+    };
+
+    using value_t = ValueType;
+
+    struct ConstructFull_t
+    {
+    };
+
+    struct ConstructEmpty_t
+    {
+    };
+
+    // @todo: a compile time check whether Index is actually lock free would be nice
+    // note: there is a way  with is_always_lock_free in c++17 (which we cannot use here)
+
+    static constexpr ConstructFull_t ConstructFull{};

```

Comment:
When you move it outside of the index queue then the index queue can be used more easily. Take a look at the std::nullopt from std::optional<T> for example. Thats how the STL uses this concept.

Reply:
I disagree here. This is used mainly internally for now and would pollute the namespace. For example nullopt_t in iceoryx is defined in the namespace cxx. It is also a bit different here I would argue. Assume I want to do something similar with another container, how would the struct be called then? ConstructFull_t would already be in use, so we would have to introduce a namespace or choose another name (which will get unwieldy fast). No, this ConstructFull_t is strongly related to the IndexQueue itself and therefore best placed in its implicit namespace.